python/ray/data/datasource/spark_datasource.py

```diff
@@ -0,0 +1,130 @@
+from ray.util.annotations import PublicAPI
+
+
+def validate_requirements():
+    from ray.util.spark.utils import get_spark_session
+
+    spark = get_spark_session()
+    if spark.conf.get(
+        "spark.databricks.pyspark.dataFrameChunk.enabled", "false"
+    ).lower() != "true":
+        raise RuntimeError(
+            "In databricks runtime, if you want to use 'ray.data.from_spark' API, "
+            "you need to set spark cluster config "
+            "'spark.databricks.pyspark.dataFrameChunk.enabled' to 'true'."
+        )
+
+    if spark.conf.get(
+        "spark.databricks.acl.dfAclsEnabled", "false"
+    ).lower() != "false":
+        raise RuntimeError(
+            "In databricks runtime, if you want to use 'ray.data.from_spark' API, "
+            "you must use an assigned mode databricks cluster."
+        )
+
+
+def _persist_dataframe_as_chunks(spark_dataframe, bytes_per_chunk):
+    try:
+        from pyspark.databricks.sql.chunk import persistDataFrameAsChunks
+    except ImportError:
+        raise RuntimeError(
+            "Current Databricks Runtime version does not support Ray SparkDatasource."
+        )
+    return persistDataFrameAsChunks(spark_dataframe, bytes_per_chunk)
+
+
+def _read_chunk_fn(chunk_ids) -> Iterator["pyarrow.Table"]:
+    if read_chunk_fn_path := os.environ.get(
+        "_RAY_DATABRICKS_FROM_SPARK_READ_CHUNK_FN_PATH"
+    ):
+        import ray.cloudpickle as pickle
+
+        # This is for testing.
+        with open(read_chunk_fn_path, "rb") as f:
+            readChunk = pickle.load(f)
+
+    else:
+        from pyspark.databricks.sql.chunk import readChunk
+
+    yield from map(readChunk, chunk_ids)
+
+
+def _unpersist_chunks(chunk_ids):
+    from pyspark.databricks.sql.chunk import unpersistChunks
+    return unpersistChunks(chunk_ids)
+
+
+@PublicAPI(stability="alpha")
+class SparkDatasource(Datasource):
+
+    def __init__(self, spark_dataframe, bytes_per_chunk):
+        self.chunk_meta_list = _persist_dataframe_as_chunks(spark_dataframe, bytes_per_chunk)
+        self.num_chunks = len(self.chunk_meta_list)
+
+        self._estimate_inmemory_data_size = sum(
+            chunk_meta.byte_count for chunk_meta in self.chunk_meta_list
+        )
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return self._estimate_inmemory_data_size
+
+    def _get_read_task(self, index, parallelism):
+        # get chunk list to be read in this task
+        chunk_index_list = list(np.array_split(range(self.num_chunks), parallelism)[index])
+
+        num_rows = sum(
+            self.chunk_meta_list[chunk_index].row_count
+            for chunk_index in chunk_index_list
+        )
+        size_bytes = sum(
+            self.chunk_meta_list[chunk_index].byte_count
+            for chunk_index in chunk_index_list
+        )
+
+        metadata = BlockMetadata(
+            num_rows=num_rows,
+            size_bytes=size_bytes,
+            schema=None,
```

Comment:
can we get schema of data?

Reply:
I think it can infer data type automatically, if you prefer explicitly setting data type, I will add it.

Reply:
explicitly setting data type would be better. But if it's too complicated, okay to leave as a followup.

Reply:
Currently spark dataframe doesn't provide public API to convert schema types to arrow types, I prefer doing it later after spark adds public API for it.

Reply:
ok, SG.